google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudLibraryProjectState.java

```diff
@@ -0,0 +1,161 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.apis;
+
+import com.google.cloud.tools.libraries.json.CloudLibrary;
+import com.google.common.collect.ImmutableSet;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.components.ProjectComponent;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Computable;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.fest.util.Maps;
+import org.fest.util.VisibleForTesting;
+import org.jetbrains.idea.maven.dom.MavenDomUtil;
+import org.jetbrains.idea.maven.dom.model.MavenDomDependency;
+import org.jetbrains.idea.maven.dom.model.MavenDomProjectModel;
+import org.jetbrains.idea.maven.project.MavenProject;
+import org.jetbrains.idea.maven.project.MavenProjectsManager;
+import org.jetbrains.idea.maven.project.MavenProjectsManager.Listener;
+
+/**
+ * A {@link ProjectComponent} that maintains, in memory, the current set of {@link CloudLibrary}
+ * that are configured on the user's project.
+ */
+public class CloudLibraryProjectState implements ProjectComponent {
+
+  private static final Logger logger = Logger.getInstance(AddCloudLibrariesAction.class);
+
+  private final Project project;
+  private Map<Module, Set<CloudLibrary>> moduleLibraryMap = Maps.newHashMap();
+  private List<CloudLibrary> allLibraries;
+
+  private CloudLibraryProjectState(Project project) {
+    this.project = project;
+  }
+
+  static CloudLibraryProjectState getInstance(Project project) {
+    return project.getComponent(CloudLibraryProjectState.class);
+  }
+
+  /**
+   * Attaches a listener to listen for changes to a module's pom.xml to request a resync of the
+   * project's managed dependencies.
+   *
+   * <p>If the project is not a maven project, then the listener simply isn't fired. If the project
+   * is not a maven project, but later becomes one, then the listener will be fired as expected and
+   * the managed libraries will be synchronized.
+   */
+  @Override
+  public void projectOpened() {
+    // todo (eshaul) handle Gradle & native projects and ensure that double syncing won't occur
+    MavenProjectsManager.getInstance(project)
+        .addManagerListener(
```

Comment:
What would happen to the listener when project is closed and then re-opened, or re-loaded - does it have a well defined behavior? Does it need to be explicitly removed?

Reply:
good question. My assumption was that the platform would be smart enough to close these "maven project" listeners when the project is closed. But i guess its dangerous to assume anything.

Ill test it out